### PR TITLE
feat: enable shadowforking

### PR DESCRIPTION
**Pull request by barnabasbusa**

### CODE REVIEW

**Comment by barnabasbusa** [comment]
example config: 
```yaml
participants:
  - el_client_type: geth
    el_client_image: ethpandaops/geth:master
    cl_client_type: lighthouse
    cl_client_image: ethpandaops/lighthouse:unstable
    count: 1
  - el_client_type: geth
    el_client_image: ethpandaops/geth:master
    cl_client_type: lodestar
    cl_client_image: ethpandaops/lodestar:unstable
network_params:
  deneb_fork_epoch: 1
  genesis_delay: 100
  network: ephemery-shadowfork
additional_services:
  - dora
snooper_enabled: true
persistent: true
```

File: README.md [hunk 1]
```diff
@@ -64,6 +64,25 @@ To mitigate these issues, you can use the `el_client_volume_size` and `cl_client
 
 For optimal performance, we recommend using a cloud provider that allows you to provision Kubernetes clusters with fast persistent storage or self hosting your own Kubernetes cluster with fast persistent storage.
 
+### Shadowforking
+In order to enable shadowfork capabilities, you can use the `network_params.network` flag. The expected value is the name of the network you want to shadowfork followed by `-shadowfork`. For example, to shadowfork the Holesky testnet, you can use the following command:
+```yaml
+...
```
Comment (h4ck3rk3y) [inline README.md:70]: Should we tell people that `persistent` is required? Or does that get auto enabled?

File: src/el/geth/geth_launcher.star [hunk 2]
```diff
@@ -99,11 +99,16 @@ def launch(
 
     network_name = (
         "devnets"
-        if launcher.network != "kurtosis"
-        and launcher.network != "ephemery"
+        if launcher.network != constants.NETWORK_NAME.kurtosis
```
Comment (h4ck3rk3y) [inline src/el/geth/geth_launcher.star:102]: `if network not in (a, b, c)`

File: src/el/geth/geth_launcher.star [hunk 3]
```diff
@@ -99,11 +99,16 @@ def launch(
...
+        and launcher.network != constants.NETWORK_NAME.ephemery
+        and launcher.network != constants.NETWORK_NAME.shadowfork
         and launcher.network not in constants.PUBLIC_NETWORKS
         else launcher.network
     )
+    if (
+        constants.NETWORK_NAME.shadowfork in network_name
+    ):  # if its a shadowfork we want proper resource allocation based on public network
+        network_name = launcher.network.split("-shadowfork")[0]
```
Comment (h4ck3rk3y) [inline src/el/geth/geth_launcher.star:111]: i wonder if we can pull out network name splitting into its own method

File: src/participant_network.star [hunk 4]
```diff
@@ -67,9 +70,115 @@ def launch_participant_network(
...
+    latest_block = ""
+    cancun_time = 0
+    prague_time = 0
+    if (
+        network_params.network == constants.NETWORK_NAME.kurtosis
+        or constants.NETWORK_NAME.shadowfork in network_params.network
+    ):
+        if (
+            constants.NETWORK_NAME.shadowfork in network_params.network
+        ):  # shadowfork requires some preparation
```
Comment (h4ck3rk3y) [inline src/participant_network.star:84]: this could use the shared name splitter; mentioned above

File: src/participant_network.star [hunk 5]
```diff
@@ -67,9 +70,115 @@ def launch_participant_network(
...
+
+                el_service_name = "el-{0}-{1}-{2}".format(
+                    index_str, el_client_type, cl_client_type
+                )
+                shadowfork_data = plan.add_service(
+                    name="shadowfork-{0}".format(el_service_name),
+                    config=ServiceConfig(
+                        image="rclone/rclone:1.55.1",
+                        cmd=[
+                            "rclone copy -P mys3:ethpandaops-ethereum-node-snapshots/"
```
Comment (h4ck3rk3y) [inline src/participant_network.star:130]: i am surprised that this runs in parallel and doesn't wait for the r clone to finish. Its nice though.

File: src/participant_network.star (hunk 4 above)
Comment (h4ck3rk3y) [inline src/participant_network.star:84]: this method is continously getting bigger; and I would like to get this refactored. It's at 640~ lines; lets refactor it into smaller parts that make sense

File: src/participant_network.star [hunk 6]
```diff
@@ -129,7 +237,7 @@ def launch_participant_network(
                 )
             else:
                 ethereum_genesis_generator_image = (
-                    "ethpandaops/ethereum-genesis-generator:3.0.0-rc.18"
+                    "ethpandaops/ethereum-genesis-generator:3.0.0-rc.19"
```
Comment (h4ck3rk3y) [inline src/participant_network.star:240]: lets put this in the constants.star please!

File: src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star [hunk 7]
```diff
@@ -55,10 +65,13 @@ def generate_el_cl_genesis_data(
         genesis_values_and_dest_filepath, "genesis-el-cl-env-file"
     )
 
+    files = {GENESIS_VALUES_PATH: genesis_generation_config_artifact_name}
+    if latest_block != "":
```
Comment (h4ck3rk3y) [inline src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star:69]: can't we just check if shadowfork_file != ""

File: src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star [hunk 8]
```diff
@@ -25,7 +26,15 @@ def generate_el_cl_genesis_data(
     capella_fork_epoch,
     deneb_fork_epoch,
     electra_fork_epoch,
+    latest_block,
 ):
+    files = {}
+    if latest_block == "":
```
Comment (h4ck3rk3y) [inline src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star:32]: perhaps the following is cleaner;

```
files = {}
shadowfork_file = ""
if latest_block != "":
    files[SHADOWFORK_FILEPATH] = latest_block
     shadowfork_file = SHADOWFORK_FILEPATH + "/shadowfork/latest_block.json"
```

note lesser indentation; and you don't need the if latest_block condition below either

File: src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star (hunk 7 above)
Comment (h4ck3rk3y) [reply src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star:69]: you won't need it if you refactor the if condition above

**Review by h4ck3rk3y — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (h4ck3rk3y) [reply README.md:70]: if persistent and node selectors are requirements for k8s; lets also mention that for now its limited to single host k8s

File: .github/tests/verkle-gen-devnet-4.yaml [hunk 9]
```diff
@@ -1,13 +1,13 @@
 participants:
   - el_client_type: geth
-    el_client_image: ethpandaops/geth:kaustinen-with-shapella-6d7b22c
+    el_client_image: ethpandaops/geth:kaustinen-with-shapella-0b110bd
```
[the remark this answers is not in the document]
Comment (barnabasbusa) [reply .github/tests/verkle-gen-devnet-4.yaml:3]: yes, verkle and shadowforks go hand in hand at this stage.

File: README.md (hunk 1 above)
Comment (barnabasbusa) [reply README.md:70]: Done

File: src/participant_network.star (hunk 6 above)
Comment (barnabasbusa) [reply src/participant_network.star:240]: done

File: src/el/geth/geth_launcher.star (hunk 3 above)
Comment (barnabasbusa) [reply src/el/geth/geth_launcher.star:115]: Done

File: src/participant_network.star (hunk 4 above)
Comment (barnabasbusa) [reply src/participant_network.star:84]: I'd focus on getting this merged, and refactor in an upcoming PR.

File: src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star (hunk 8 above)
Comment (barnabasbusa) [reply src/prelaunch_data_generator/el_cl_genesis/el_cl_genesis_generator.star:34]: done

File: src/el/besu/besu_launcher.star [hunk 10]
```diff
@@ -208,7 +208,10 @@ def get_config(
         "--metrics-host=0.0.0.0",
         "--metrics-port={0}".format(METRICS_PORT_NUM),
     ]
-    if network not in constants.PUBLIC_NETWORKS:
+    if (
```
[the remark this answers is not in the document]
Comment (barnabasbusa) [reply src/el/besu/besu_launcher.star:205]: its not the same logic. 

I want to use not in some cases and != in other cases. Its not the same thing.

File: src/participant_network.star (hunk 4 above)
Comment (barnabasbusa) [reply src/participant_network.star:84]: tracking it here: https://github.com/kurtosis-tech/ethereum-package/issues/492

File: src/participant_network.star [hunk 11]
```diff
@@ -98,7 +102,7 @@ def launch_participant_network(
                 run="mkdir -p /shadowfork && \
                     curl -o /shadowfork/latest_block.json https://ethpandaops-ethereum-node-snapshots.ams3.digitaloceanspaces.com/"
                 + base_network
-                + "/geth/latest/latest_snapshot_block.json",
+                + "/geth/" + shadowfork_block + "/_snapshot_eth_getBlockByNumber.json",
```
Comment (barnabasbusa) [inline src/participant_network.star:105]: this is not what Rafael called the json files. Maybe we can rename this to be just block.json?

File: src/participant_network.star [hunk 12]
```diff
@@ -130,9 +134,14 @@ def launch_participant_network(
                             + base_network
                             + "/"
                             + el_client_type
-                            + "/latest /data/"
+                            + "/" + shadowfork_block + " /data/"
+                            + el_client_type
+                            + "/snapshot.tar.zst | tar -I zstd -xvf - -C /data/"
```
Comment (barnabasbusa) [inline src/participant_network.star]: zstd might need to be installed

**Pull request merged**